src/main/java/org/opensearch/flowframework/util/ParseUtils.java

```diff
@@ -364,10 +373,19 @@ private static Object conditionallySubstitute(Object value, Map<String, Workflow
             m.appendTail(result);
             value = result.toString();
 
-            // Replace all params if present
-            for (Entry<String, String> e : params.entrySet()) {
-                String regex = "\\$\\{\\{\\s*" + Pattern.quote(e.getKey()) + "\\s*\\}\\}";
-                value = ((String) value).replaceAll(regex, e.getValue());
+            if (params != null) {
+                for (Map.Entry<String, String> e : params.entrySet()) {
+                    String regex = "\\$\\{\\{\\s*" + Pattern.quote(e.getKey()) + "\\s*\\}\\}";
+                    String replacement = e.getValue();
+
+                    // Special handling for JSON strings that contain placeholders (connectors action)
+                    replacement = Matcher.quoteReplacement(replacement.replace("\"", "\\\""));
+
+                    // Use Pattern.compile().matcher() to avoid issues with replaceAll's direct pattern compilation
+                    Pattern pattern = Pattern.compile(regex);
+                    Matcher matcher = pattern.matcher((String) value);
+                    value = matcher.replaceAll(replacement);
```

Comment:
Instead of pattern can we use StringBuilder here. Something like

```java
new StringBuilder(modifiedValue.toString().replaceAll(regex, replacement));
```

Reply:
I was getting a bug with replaceAll that it didn't diffrentiate a more complicated substiutitin, for example when the string to be substituted in is: "{ \"texts\": ${parameters.texts}, \"truncate\": \"${parameters.truncate}\", \"model\": \"${parameters.model}\", \"input_type\": \"${parameters.input_type}\" }"

it would get confused with the other ${}, I can try this again though after I made changes after

Reply:
Declare a StringBuilder like

```
StringBuilder modifiedValue = new StringBuilder(value);
```

Replace it like

```
modifiedValue = new StringBuilder(modifiedValue.toString().replaceAll(regex, replacement));
```

then your final value would be `value=modifiedValue.toString()` because you will convert StringBuilder to String here

Reply:
I see, i actually tested it out now and it works like this even: 
`value = ((String) value).replaceAll(regex, replacement);`

Reply:
Didn't work with `StringBuilder`?

Reply:
Works same without it too though cause anyways I want to convert in the end to a string